zarr/storage.py

```diff
@@ -1877,6 +1879,365 @@ def __delitem__(self, key):
+        keys = list(map(operator.itemgetter(0), keys))
+        return keys
+
+    def getsize(self, path=None):
+        path = normalize_storage_path(path)
+        size = self.cursor.execute(
+            '''
+            SELECT COALESCE(SUM(LENGTH(v)), 0) FROM {t}
+            WHERE k LIKE "{p}%" AND
+                  0 == INSTR(LTRIM(SUBSTR(k, LENGTH("{p}") + 1), "/"), "/")
+            '''.format(
+                t=self.table, p=path
+            )
+        )
+        for s, in size:
+            return s
+
+    def rename(self, src_path, dst_path):
+        src_path = normalize_storage_path(src_path)
+        dst_path = normalize_storage_path(dst_path)
+
+        self.cursor.executescript(
+            '''
+            BEGIN TRANSACTION;
+                CREATE TEMPORARY TABLE _{t}_{u} AS
+                SELECT LTRIM("{dp}" || "/" || mk, "/"), mv FROM (
+                    SELECT LTRIM(SUBSTR(k, LENGTH("{sp}") + 1), "/") AS mk,
+                           v AS mv
+                    FROM {t} WHERE k LIKE "{sp}%"
+                );
+                DELETE FROM {t} WHERE k LIKE "{sp}%";
+                REPLACE INTO {t} SELECT * FROM _{t}_{u};
+                DROP TABLE _{t}_{u};
+            COMMIT TRANSACTION;
+            '''.format(
+                t=self.table, u=uuid.uuid4().hex, sp=src_path, dp=dst_path
+            )
+        )
+
+    def rmdir(self, path=None):
+        path = normalize_storage_path(path)
+        if path:
+            self.cursor.execute(
+                '''
+                DELETE FROM {t} WHERE k LIKE "{p}_%"
+                '''.format(
+                    t=self.table, p=path
+                )
+            )
+        else:
+            self.clear()
+
+    def clear(self):
+        self.cursor.executescript('''
+            BEGIN TRANSACTION;
+                DROP TABLE {t};
+                CREATE TABLE {t}(k TEXT PRIMARY KEY, v BLOB);
+            COMMIT TRANSACTION;
+        '''.format(t=self.table))
+
+
+class MongoDBStore(MutableMapping):
+    """Storage class using MongoDB.
+
+    Parameters
+    ----------
+    database : string
+        Name of database
+    collection : string
+        Name of collection
+    **kwargs
+        Keyword arguments passed through to the `pymongo.MongoClient` function.
+    """
+
+    _key = 'key'
+    _value = 'value'
+
+    def __init__(self, database='mongodb_zarr', collection='zarr_collection',
+                 **kwargs):
+        import pymongo
+
+        self._database = database
+        self._collection = collection
+        self._kwargs = kwargs
+
+        self.client = pymongo.MongoClient(**self._kwargs)
+        self.db = self.client.get_database(self._database)
+        self.collection = self.db.get_collection(self._collection)
+
+    def __getitem__(self, key):
+        doc = self.collection.find_one({self._key: key})
+
+        if doc is None:
+            raise KeyError(key)
+        else:
+            return binary_type(doc[self._value])
```

Comment:
Interesting. Wonder if `ensure_bytes` would work here as well. 🤔

Reply:
Out of interest, what type of object is `doc[self._value]`?

Reply:
AFAICT it's a `bytes` object. Am curious to see if we may have already fixed this with other changes and this can be dropped. Have added PR ( https://github.com/zarr-developers/zarr/pull/401 ) to give this a try.